Add fallback routing

rynowak opened the pull request:
Addresses #7928

Adds the idea of a *fallback* route to routing - as an opinionated way to set things up. 

Adds extension methods for:
- *fallback to generic delegate*
- *fallback to static file*
- *fallback to controller/page*

rynowak committed:
Add file/non-file and generic fallback

Adds new constraints for checking if a route value is a file or not.

Added a new set of builder methods that specify what it means to be a
'fallback'. This is really similar to what the older SPA fallback routes
do, but this is lower in the stack and directly integrated with
endpoints.

rynowak committed:
Add fallback routing to file

This is a new routing feature that integrates static files to serve a
static file when routing doesn't match anything else.

This is a scenario that's covered by SPA services today, but given the
improvements to routing it makes much more sense to move lower in the
stack.

rynowak committed:
Update SpaDefaultPageMiddleware for endpoints

This change makes the SpaDefaultPageMiddleware noop when an endpoint has
been chosen. This is a problem today because this usually runs after
routing, but its always terminal.

So even if endpoint routing selected something, this would still serve
the default file. We're adding more things that fill this niche, but it
seems like a good idea to fix the existing stuff.

rynowak committed:
Generalize Action Selection logic

Allows us to use the "select action via route values" logic for
endpoints.

rynowak committed:
Add fallback routing for controllers and pages

rynowak commented:
It's probably a good idea to look at individual commits because the changes are all over.

rynowak commented on src/Http/Routing/src/Builder/FallbackEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <see cref="MapFallback(IEndpointRouteBuilder, RequestDelegate)"/> registers an endpoint using the pattern
+        /// <c>{*path:nonfile}</c>. The order of the registered endpoint will be <c>int.MaxValue</c>.
+        /// </para>
+        /// </remarks>
```
The main scenario here is carried over from spa services routing things, and represents the idea that you want the static files middleware to be responsible for things that **look like** files, and you want routing to handle things that **don't**. It's an opinionated way to do a bunch of useful things.

rynowak replied:
The methods here are layered on top of existing primitives (ordering, constraints).

rynowak commented on src/Middleware/StaticFiles/ref/Microsoft.AspNetCore.StaticFiles.csproj:
```diff
@@ -7,6 +7,7 @@
     <Compile Include="Microsoft.AspNetCore.StaticFiles.netcoreapp3.0.cs" />
     <Reference Include="Microsoft.AspNetCore.Hosting.Abstractions"  />
     <Reference Include="Microsoft.AspNetCore.Http.Extensions"  />
+    <Reference Include="Microsoft.AspNetCore.Routing"  />
```
@Tratcher - thoughts? What's the right place for these extension methods to live.

Tratcher replied:
See my comments elsewhere. It's not clear you need to bring in the middleware in order to serve a single file.

rynowak replied:
Discussed offline. We're going to go with this for now and evaluate doing some deeper integration between these components.

rynowak commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,223 @@
+            this IEndpointRouteBuilder builder,
+            string pattern,
+            string filePath,
+            StaticFileOptions options)
```
@SteveSandersonMS - I carried forward the options from the spa services stuff because it seemed like there was a lot of complexity there that I don't totally get.

I'm interested in whether you think these things can be used to solve the same set of problems or not.

SteveSandersonMS replied:
I would strongly expect that `MapFallbackToFile` could replace the bulk of `UseSpaStaticFiles`.

The `StaticFileOptions` are needed if you want to set custom response headers (e.g., for caching policy), which people do.

Other complications in `UseSpaStaticFiles` are to do with things like "at dev time I proxy requests to a SPA dev server, but in production is serve prebuilt static files from my contentroot".

rynowak replied:
OK cool. That was what I was hoping for. I'm confident that we can build more flavours of this if needed.

rynowak commented on src/Middleware/SpaServices.Extensions/src/SpaDefaultPageMiddleware.cs:
```diff
@@ -24,6 +25,12 @@ public static void Attach(ISpaBuilder spaBuilder)
+                {
+                    return next();
+                }
+
```
I coulnd't find any tests. #YOLO

SteveSandersonMS replied:
Historically it was E2E tests, but they got removed. So yeah.

rynowak commented on src/Http/Routing/src/DataSourceDependentCache.cs:
```diff
@@ -8,6 +8,8 @@
 {
+    // FYI: This class is also linked into MVC. If you make changes to the API you will
+    // also need to change MVC's usage.
     internal sealed class DataSourceDependentCache<T> : IDisposable where T : class
```
This commit has a bunch of random cleanup because I'm trying to share the core data structure used in action selection in a few places.

rynowak commented on src/Mvc/Mvc.Core/src/Infrastructure/ActionSelectionTable.cs:
```diff
@@ -0,0 +1,155 @@
+                }
+            }
+
+            return new ActionSelectionTable<T>(version, routeKeys.ToArray(), ordinalEntries, ordinalIgnoreCaseEntries);
```
This is all existing code I ripped out of action selector.

rynowak commented on src/Mvc/Mvc.Core/src/Infrastructure/StringArrayComparer.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace Microsoft.AspNetCore.Mvc.Infrastructure
+{
+    internal class StringArrayComparer : IEqualityComparer<string[]>
```
This is all existing code I ripped out of action selector.

rynowak commented on src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,113 @@
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class DynamicControllerEndpointMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
```
There are a few different design approaches I could have taken with MVC:

1. Add more nodes to the DFA stuff by doing an application model thing
2. Use the fallback endpoint's delegate to *chain* to an MVC endpoint
3. Use a matcher policy to substitute in another endpoint dynamically.

I rejected 1 because I don't really want this new routing feature to hang off of MVC options/conventions. Another idea that I could do would be to build this into the data sources themselves, which might be fine?

I rejected 2 because it has problems with things like authZ. You want the policies applied to the endpoint to be applied by middleware and that's only possible if endpoint is correct when middleware run.

I did 3 because I think we're **going to need this functionality**. I have another work item planned to implement features that people will use to port their `IRouter` code forward. What's enabled by this change is 100% needed for that.

-----

I'm looking for input on this because there are a few real tradeoffs here in terms of what's possible and how future enhancements will play out.

rynowak replied:
Also @davidfowl I summon you.

javiercn replied:
I believe you are right (as usual) in this.
1 would work but I understand why you don't want it.
2 makes sense.
3. I think I get it.

As someone not that familiar with this part of the codebase it would be great to get a more explicit pros/cons list. As it seems you clearly prefer 3. Is there any drawback to it?

rynowak replied:
@JamesNK @pranavkm @davidfowl

pranavkm replied:
>  I have another work item planned to implement features that people will use to port their IRouter code forward. 

What sort of routing scenario would this help in?

rynowak replied:
This is the typical CMS thing. Write an `IRouter` that treats the URL as a *slug* and then look that up in the database and put in the route values that go to the desired action.

rynowak commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,113 @@
+                {
+                    continue;
+                }
+
```
Note to myself, this isn't done. I need to use the pageloader inside the selector.

rynowak commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,113 @@
+            }
+        }
+    }
+}
```
There's a certain amount of de-duplication that could be done between these two pages (controller and pages). 

I'm not incredibly keen because to a certain extent these code paths need to each have their own types for everything because of DI/metadata. Open to suggestions/comments 😆

javiercn replied:
Make it generic based on the selector and the metadata type :trollface:

rynowak replied:
Resolving this comment. I feel like I solved this for the most part by factoring more code into the action selection table class.

rynowak commented on src/Mvc/samples/MvcSandbox/Startup.cs:
```diff
@@ -68,6 +68,8 @@ public void Configure(IApplicationBuilder app)
                 builder.MapControllers();
                 builder.MapRazorPages();
+
+                builder.MapFallbackToController("Index", "Home");
```
example usage.

javiercn replied:
This is neat.
Also you might have forgotten to register the page related stuff on DI

rynowak replied:
yeah, I haven't written tests yet for the MVC cases, and I've only manually tested the controller case.

rynowak commented on src/Mvc/Mvc.RazorPages/src/Builder/RazorPagesEndpointRouteBuilderExtensions.cs:
```diff
@@ -31,6 +32,245 @@ public static IEndpointConventionBuilder MapRazorPages(this IEndpointRouteBuilde
+                // MVC registers a policy that looks for this metadata.
+                b.Metadata.Add(CreateDynamicPageMetadata(page, area));
+            });
+        }
```
I'm really please how much powerful stuff can be built here with the building blocks we have available (ordering, constraints, policies, endpoints). None of these features required new features, it's all existing extensibility. @JamesNK - you're the best around.

javiercn commented on src/Http/Routing/test/UnitTests/Constraints/FIleNameRouteConstraintTest.cs:
```diff
@@ -0,0 +1,99 @@
+                    "/////hello.x",
+                    "a/b/c/d.e",
+                    "a/b./.c/d.e",
+                };
```
I would also add a leading dot combinations like `.gitignore`

SteveSandersonMS replied:
On balance I agree. Leading-dotted things probably are filenames, and are not likely to be desired as part of the client-side URL space. It's a weird scenario, though.

rynowak replied:
yeah on reflection, I agree. I'll update this.

javiercn commented on src/Http/Routing/test/UnitTests/Constraints/FIleNameRouteConstraintTest.cs:
```diff
@@ -0,0 +1,99 @@
+                    string.Empty,
+                    "/",
+                    ".",
+                    ".a",
```
Seems like you are not considering this a file, what's the reason?

javiercn commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,223 @@
+            string filePath,
+            StaticFileOptions options = null)
+        {
+            var app = builder.CreateApplicationBuilder();
```
This is neat, I didn't know you could do this.

rynowak replied:
@davidfowl hates this, but he's wrong. It's the bees knees.

rynowak replied:
btw the technical term is routerware

javiercn commented on src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,113 @@
+            _selector = selector;
+        }
+
+        public override int Order => int.MinValue + 100;
```
Suggestion: Could we make all these offsets consts into an internal shared class that everybody links so that we can see the ordering in one place? It might sounds stupid, but it's really hard to think about the ordering of the system when it's distributed across classes. I also understand that this is kind of unrelated and a nit.

rynowak replied:
It's kinda a mess no matter what. I have a few ideas.

javiercn commented on src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointSelector.cs:
```diff
@@ -0,0 +1,72 @@
+                throw new ArgumentNullException(nameof(values));
+            }
+
+            var state = CurrentState;
```
nit: table is much better than state IMHO.

rynowak replied:
Fowler hates this so much, but he's wrong.

rynowak replied:
meh.

javiercn commented on src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointSelector.cs:
```diff
@@ -0,0 +1,72 @@
+            var extractedValues = new string[keys.Length];
+            for (var i = 0; i < keys.Length; i++)
+            {
+                values.TryGetValue(keys[i], out var obj);
```
Is it guaranteed that at this point this data is going to be here? I think that deserves a comment

rynowak replied:
I'm not sure what you are concerned about. It's `TryGetValue`?

javiercn replied:
Yeah, keep in mind that I'm not 100% familiar with the abstractions, so I don't have a full understanding of things

javiercn commented on src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,113 @@
+                }
+
+                var matchedValues = candidates[i].Values;
+                var endpoints = await _selector.SelectEndpointsAsync(metadata.Values);
```
Does select throw when multiple endpoints get picked up?

rynowak replied:
No, the documentation says that it's implementation defined. The reason why I made this implementation defined ordering instead of throwing is that we will have multiple matches in a trivial case - pages that have the filename `Index.cshtml`. This produces two endpoints. 

-----

We could change this depending on scenarios and feedback. Here are the concerns involved, I know @JamesNK and @davidfowl will be interested as well.

1. If you want this to throw - it breaks a common use case
2. If you want this to be implementation defined it's really simple to implement and explain, but the capabilities are pretty limited
3. If you want this to **allow** returning multiple endpoints, and have the rest of the system try to **disambiguate** them for you, then it's a very powerful and confusing system.

----

So let's assume that you want to be able to return multiple endpoints (3) and explore it a bit. Let's consider a basic case, like "I want to return two endpoints, and have HTTP method disambiguation select the best match".

Well, The *dynamic endpoint pattern* allows you to **RUN CODE** to replace the dynamic endpoint with the single endpoint of your choosing - **AFTER** the route pattern has been matched.

So I really need to add multiple endpoints but I can't because ....

This is a intentional limitation of `CandidateSet`. It's a data structure that's highly optimized for a small number of entries, and it doesn't allow you to insert items. We could change this, and it would be relatively inefficient everywhere you use it (relative to something like techempower).

Cool, so now I can insert multiple endpoints. The next problem is that ...

The matcher policy that you write to replace an endpoint runs in a pipeline of matcher policies. ie: some policies might run before, some might run after.  Well, if you want the system to disambiguate for you then you really are concerned about which policies run before you and which after. Policies that ran before you will see the dynamic endpoint, policies that run after you will see the replacement.

The next issue is that we have policies that bake themselves into the DFA. We've generally written our policies to use the DFA instead of running in the pipeline of matcher policies (HTTP method, Consumes, Hostname). This is like the difference between compiled (DFA) and interpreted (matcher policy). The DFA policy implementations become part of the core data structure of routing, but they cannot do this for a dynamic endpoint, because the metadata of that endpoint are unknown. We will need to make it possible for all of our policies to run in DFA model or `IEndpointSelectorPolicy` mode.

Now something like this could be possible.

If we're going to end up here - ie we need these features, then I think I would step back and make this a core feature of DFA matcher rather than a thing built on top.

rynowak replied:
I like option 2 (until we know more or discuss more) because while it might not be convenient, it's not blocking either. Someone who wants to extend routing is taking on some *hard problems* - I don't really assume we can solve them for everyone.

javiercn commented on src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,113 @@
+                        "{ " + string.Join(", ", metadata.Values.Select(kvp => $"{kvp.Key}: {kvp.Value}")) + " }.");
+                }
+
+                var replacement = endpoints[0];
```
This seems sketchy if the above call doesn't throw and we can't guarantee that this always results in a single endpoint being selected. I might not understand all the details of how this works, but it feels like when you had two matching actions.

rynowak replied:
Resolving this comment since I replied to this concern above.

javiercn reviewed (commented):
These changes look great!
Pretty much inline with what I thought it would look like.

I'm assuming that you use order to prevent legit requests to things like `/api/values` to match this if there is already an action for that thing. I think we could benefit from seeing a more 'realistic' example, as in I want to prevent '/api/*' values from matching to prevent a mistake in calling an API to result in a 200. 

How would that look like in Startup?

javiercn commented:
I did a first pass, let me know when you are ready for review

SteveSandersonMS commented on src/Http/Routing/src/Builder/FallbackEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,90 @@
+    {
+        /// <summary>
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will match
+        /// requests for non-filenames with a very low priority.
```
`a very low priority` sounds a bit imprecise. Since it's `Order = int.MaxValue`, how about `the lowest possible priority`?

javiercn replied:
Lower number > priority. 0 > 1 if that makes sense. It's been how routing has worked since forever

SteveSandersonMS replied:
All I'm commenting on here is the phrasing used in the XML docs. `a very low priority` is an imprecise statement (just *how* low is the priority?), whereas `the lowest possible priority` tells you something exact.

rynowak replied:
sure

SteveSandersonMS commented on src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,145 @@
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// This constraint can be used with a low priority route to disambiguate for requests for
```
The level of priority in the route doesn't really matter, does it? The constraint only deals with what the URL looks like, not whether it's useful to prioritize it above or below other routes.

rynowak replied:
I described it here to try and capture the main use case in the docs. I think it's valuable to capture how we expect something to used in a meaningful way. Let me know if you have ideas to improve this.

SteveSandersonMS commented on src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,145 @@
+    /// a legal file name for the current operating system.
+    /// </para>
+    /// <para>
+    /// <list type="bullet">  
```
I'm interested to know how all this content surfaces to users. It gets injected into the auto-generated API documentation web pages, does it?

rynowak replied:
That's the idea at least. I don't have an example handy.

SteveSandersonMS commented on src/Http/Routing/test/UnitTests/Constraints/FIleNameRouteConstraintTest.cs:
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Typo in filename: `FIleNameRouteConstraintTest.cs` should be `FileNameRouteConstraintTest.cs`

SteveSandersonMS commented on src/Mvc/Mvc.Core/src/Infrastructure/ActionSelectionTable.cs:
```diff
@@ -0,0 +1,155 @@
+{
+    // Common infrastructure for things that look up actions by route values.
+    //
+    // The ActionSelectionTablee stores a mapping of route-values -> items for each known set of
```
```suggestion
    // The ActionSelectionTable stores a mapping of route-values -> items for each known set of
```

SteveSandersonMS commented on src/Mvc/Mvc.RazorPages/src/Builder/RazorPagesEndpointRouteBuilderExtensions.cs:
```diff
@@ -31,6 +32,245 @@ public static IEndpointConventionBuilder MapRazorPages(this IEndpointRouteBuilde
+        /// </remarks>
+        public static void MapFallbackToPage(
+            this IEndpointRouteBuilder routes,
+            string pattern,
```
Is there a bug here? `pattern` seems to be unused (besides the nullcheck).

SteveSandersonMS commented on src/Mvc/Mvc.RazorPages/src/Builder/RazorPagesEndpointRouteBuilderExtensions.cs:
```diff
@@ -31,6 +32,245 @@ public static IEndpointConventionBuilder MapRazorPages(this IEndpointRouteBuilde
+        /// <summary>
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will match
+        /// requests for non-filenames with a very low priority. The request will be routed to a page endpoint that
+        /// matches <paramref name="page"/>.
```
I'm wondering whether it might be better *not* to have the various `MapFallbackTo*` overloads that don't take a `pattern`. Those ones all "match requests for non-filenames", but TBH that feels like a pretty arbitrary specialization. I know it matches our current use cases, but it feels like a separate concern from "mapping a fallback route".

It's not like developers will be writing lots of calls to `MapFallbackTo*`. Chances are, they will have exactly one such call in their whole application, and in most of those cases it was put there for them by the template anyway.

If the template produced more explicit calls, e.g.,

    routes.MapFallbackToPage("Index.cshtml", "{*path:nonfile}");

... then it would be more obvious that the `nonfile` constraint is being applied, plus how to remove it if you don't want it.

rynowak replied:
I mostly agree and it would simplify the docs. I tend to prefer these kinds of things being explicit anyway. What I don't want to is have to put these simpler overloads back as the result of a template review.

rynowak replied:
I mocked this up here: https://gist.github.com/rynowak/ab9305b5f1dbb2d918e568b4129a8829

I can't really see us putting this route in a template because it's a lot to explain. I'm going to leave the simpler overloads in place for now, and we can review and decide later.

SteveSandersonMS replied:
Alright, I agree.

SteveSandersonMS reviewed (commented):
The new APIs look super - very clean and to the point.

I can't really comment about the internal implementation details because I have no up-to-date background in MVC action selection, endpoint routing policies, and so on. It looks believable though.

anurse commented:
FYI @rynowak, you should rebase on `master` in your next iteration. The NodeServices tests are failing in Helix right now and they are skipped in `master`. It's not critical since there are other failing tests (since this is a Draft PR).

JamesNK commented on src/Http/Routing/src/Builder/FallbackEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,90 @@
+                throw new ArgumentNullException(nameof(requestDelegate));
+            }
+
+            var coventionBuilder = builder.Map(pattern, "Fallback", requestDelegate);
```
covention -> convention

JamesNK commented on src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,145 @@
+namespace Microsoft.AspNetCore.Routing.Constraints
+{
+    /// <summary>
+    /// Constrains a route parameter to represent only filename values.
```
IMO it worth commenting that this route constraint only checks whether the parameter is formatted like a file name, and doesn't check whether the file exists on disk. That was my first thought when I saw the class name.

rynowak replied:
Sure, I agree. I mentioned it for `MapFallback` docs, but not here.

JamesNK commented on src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,145 @@
+            return false;
+        }
+
+        internal static bool IsFileName(ReadOnlySpan<char> value)
```
Certain characters are invalid in a file name, e.g. `:`, `&`. What about checking for those? To complicate things, invalid file name characters are probably different between operating systems.

Whether this constraint does or doesn't check for them then it would be worth commenting on in the XML docs.

rynowak replied:
hmm, I thought I did document that. I'll double check. 

It's definitely intended that we don't validate these. 1 - it's platform specific, 2 - it's not aligned with the purpose of this

JamesNK commented on src/Middleware/SpaServices.Extensions/src/StaticFiles/SpaStaticFilesExtensions.cs:
```diff
@@ -110,8 +111,8 @@ public static void UseSpaStaticFiles(this IApplicationBuilder applicationBuilder
-            
-            app.UseStaticFiles(staticFileOptions);
+            // We should only serve files when there is no endpoint.
+            app.MapWhen(context => context.GetEndpoint() == null, a => app.UseStaticFiles(staticFileOptions));
```
This is funky, but should it be added as a setting to staticFileOptions? That would be more user friendly/discoverable.

rynowak replied:
Maybe.... this isn't incredibly common to do. All of our guidance for static files is to put it ahead of routing/mvc. This would only come up in a very specialized use case like this.

JamesNK commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,223 @@
+        }
+
+        /// <summary>
+        /// Registers a fallback <see cref="RouteEndpoint"/> that will serve the provided static file
```
nit: fullstop

JamesNK commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,223 @@
+    public static class StaticFilesEndpointRouteBuilderExtensions
+    {
+        /// <summary>
+        /// Registers a fallback <see cref="RouteEndpoint"/> that will serve the provided static file
```
nit: fullstop

JamesNK commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,223 @@
+        /// <remarks>
+        /// <para>
+        /// <see cref="MapFallbackToFile(IEndpointRouteBuilder, string)"/> is intended to handle cases where URL path of
+        /// the request does not contain a filename, and no other endpoint has matched. This is convenient for routing
```
"does not contain a filename"

Should this be does?

rynowak replied:
nope - it's correct.

This scenario is:
1. Requests for a file get handled by static file middleware
2. Requests for a non-file get re-routed to the SPA (which happens to a static file in this case)
3. Requests for a file that don't exist on disk result in a 404.

JamesNK commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,223 @@
+        }
+
+        /// <summary>
+        /// Registers a fallback <see cref="RouteEndpoint"/> that will serve the provided static file
```
nit: fullstop. Repeat as needed.

JamesNK commented on src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs:
```diff
@@ -152,6 +153,296 @@ public static IEndpointConventionBuilder MapDefaultControllerRoute(this IEndpoin
 
+        /// <summary>
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will match
+        /// requests for non-filenames with a very low priority. The request will be routed to a controller endpoint that
```
nit: filenames is two words in other comments

JamesNK commented on src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs:
```diff
@@ -152,6 +153,296 @@ public static IEndpointConventionBuilder MapDefaultControllerRoute(this IEndpoin
+        /// <remarks>
+        /// <para>
+        /// <see cref="MapFallbackToController(IEndpointRouteBuilder, string, string)"/> is intended to handle cases where URL path of
+        /// the request does not contain a filename, and no other endpoint has matched. This is convenient for routing
```
filename again. Pick one to be consistent with one.

JamesNK commented on src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs:
```diff
@@ -152,6 +153,296 @@ public static IEndpointConventionBuilder MapDefaultControllerRoute(this IEndpoin
+
+            // Maps a fallback endpoint with an empty delegate. This is OK because
+            // we don't expect the delegate to run. 
+            routes.MapFallback(context => Task.CompletedTask).Add(b =>
```
nit: Have a shared EmptyRequestDelegate property for `context => Task.CompletedTask` and reuse

rynowak replied:
I haven't done that because there really isn't a good place for that to live that wants to be public. It's static anyway so this is already *free*

JamesNK commented on src/Mvc/Mvc.Core/src/Infrastructure/ActionSelectionTable.cs:
```diff
@@ -0,0 +1,155 @@
+                actions.Items.Where(a => a.AttributeRouteInfo == null),
+
+                a => a.RouteValues.Keys,
+                (a, key) =>
```
nit: use named parameters with these args to make it easier to understand what each is for

JamesNK commented on src/Mvc/Mvc.Core/src/Microsoft.AspNetCore.Mvc.Core.csproj:
```diff
@@ -27,6 +27,10 @@ Microsoft.AspNetCore.Mvc.RouteAttribute</Description>
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Include="..\..\..\Http\Routing\src\DataSourceDependentCache.cs" Link="Routing\DataSourceDependentCache.cs" />
```
Move the file to a Shared folder?

rynowak replied:
*eh* - I don't really free like this helps.

JamesNK replied:
I don't like a project in a different solution referencing a file. It will be easy for someone to change the API and only discover another solution is referencing it when the build fails.

rynowak replied:
I don't really like it either. I left a comment in `DataSourceDependentCache` so that it's more obvious. I don't see how moving the file on disk helps.

JamesNK reviewed (commented):
I got up to Microsoft.AspNetCore.Mvc.Core.csproj. Will review the rest later

rynowak commented:
OK this is ready for review.

Tratcher commented on src/Middleware/SpaServices.Extensions/src/StaticFiles/SpaStaticFilesExtensions.cs:
```diff
@@ -110,8 +111,8 @@ public static void UseSpaStaticFiles(this IApplicationBuilder applicationBuilder
-            
-            app.UseStaticFiles(staticFileOptions);
+            // We should only serve files when there is no endpoint.
+            app.MapWhen(context => context.GetEndpoint() == null, a => app.UseStaticFiles(staticFileOptions));
```
Worlds collide. Shouldn't this check go inside of static files so it doesn't hijack requests that already have routes? Or do we need to start work on static file based endpoints?

rynowak replied:
I'm happy to send another PR that puts this inside of static files if you think that's the right thing to do.

rynowak replied:
> Or do we need to start work on static file based endpoints so you can remove this middleware entirely?

I thought this was that? What part do you think we should make more first class?

Tratcher replied:
Yes, please update StaticFiles.

rynowak replied:
Will be in the next iteration.

Tratcher commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+
+            if (options == null)
+            {
+                app.UseStaticFiles();
```
The main things you get from the static file middleware are an IFileProvider, content-type resolution, and conditionals support (if-modified-since...).

The file provider can be assumed (IWebHostEnvironment.WebRootFileProvider), content-types have defaults or the user can specify the content-type if they're hard-coding the path, and it's not clear conditionals would be useful for fallback scenarios.

I don't think you need the full middleware to handle this. Directly calling one of the HttpResponse.[SendFileAsync](https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http.Extensions/src/SendFileResponseExtensions.cs) overloads seems adequate for this scenario. 

E.g.
- `MapFallbackToFile(... filePath)` content-types is resolved from a default set and the file is resolved from IWebHostEnvironment.WebRootFileProvider
- `MapFallbackToFile(... filePath, contentType)` is resolved from IWebHostEnvironment.WebRootFileProvider
- `MapFallbackToFile(... filePath, contentType fileProvider)` is resolved from the given provider
- `MapFallbackToFile(... iFileInfo, contentType)` sends the given file directly

Yet again, if we rewrote static files as endpoints this wouldn't be an issue, you'd new up a pre-resolved static file endpoint.

rynowak replied:
@SteveSandersonMS can elaborate more than I can on the scenarios. The existing SPA stuff and the people that use it use quite a lot of those features. 

In particular caching headers, and multiple file providers have come up in our discussions. It would be really nice if as part of this experience users could configure these things in one place and have them work for both static files and this. 

-----

I'm still open to the idea that this would be something different from the static files middleware, but what's the drawback of using the middleware?

SteveSandersonMS replied:
The strongest reason we had to use `UseStaticFiles` in SPA-land was to get all the ETag stuff for free. You really have to support this, and `HttpResponse.SendFileAsync` won't give it to you.

I agree it would be nice if the relevant parts of `UseStaticFiles` were factored out, but I also feel (like @rynowak) that there's no obvious drawback in just reusing `UseStaticFiles` here as-is.

rynowak replied:
Discussed offline and we're going to keep this as-is for now. We're meeting after preview 4 to design another approach for static files in general.

pranavkm commented on src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,146 @@
+
+            for (var i = dotIndex + 1; i < value.Length; i++)
+            {
+                if (value[i] != '.')
```
I don't super follow this. Wouldn't this prevent `foo.bar.baz` which the documentation says is allowed?

Tratcher replied:
It only needs one non-period character after the first period.

pranavkm replied:
Hmm, so if I understand correctly, it's trying to disallow files that look like `.....`, while allowing files that look like `..a.b...c.`? This needs some comments.

rynowak replied:
My goal is to allow `foo.bar` and `.foobar` but not nonesense like `..`.

pranavkm commented on src/Http/Routing/src/Constraints/NonFileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,114 @@
+    /// </list>
+    /// </para>
+    /// </remarks>
+    public class NonFileNameRouteConstraint : IRouteConstraint
```
Can you add a `//` comment that says this document needs to be kept in sync with `FileNameRouteConstraint`?

rynowak replied:
Added to the IsFileName method

pranavkm commented on src/Http/Routing/src/Constraints/NonFileNameRouteConstraint.cs:
```diff
@@ -0,0 +1,114 @@
+            if (values.TryGetValue(routeKey, out var obj) && obj != null)
+            {
+                var value = Convert.ToString(obj, CultureInfo.InvariantCulture);
+                return !FileNameRouteConstraint.IsFileName(value);
```
`FileNameHelper.IsFileName` 🚎

rynowak replied:
nope.

pranavkm commented on src/Http/Routing/test/UnitTests/Constraints/FIleNameRouteConstraintTest.cs:
```diff
@@ -0,0 +1,100 @@
+
+        [Theory]
+        [MemberData(nameof(FileNameData))]
+        public void Match_RouteValue_IsFileName(object value)
```
Could you also add a unit test that specifically tests `FileNameRouteConstraint.IsFileName`? 

```C#
[Theory]
[MemberData(nameof(FileNameData))]
public void FileNameConstraint_IsFileNameReturnsTrue(string fileName)
{
   // Act
  var result = FileNameConstraint.IsFileName(fileName);

   // Assert
   Assert.True(result);
}
```

The method is internal and we don't necessarily need the ceremony of routing to verify that method.

pranavkm replied:
Same deal with the NonFileNameData

rynowak replied:
I don't really get why testing the same thing twice will help.

pranavkm commented on src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,219 @@
+
+        /// <summary>
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will match
+        /// requests for non-filenames with the lowest possible priority. The request will be routed a
```
will be routed ~~a~~ to the StaticFileMiddleware

pranavkm commented on src/Mvc/Mvc.RazorPages/src/Builder/RazorPagesEndpointRouteBuilderExtensions.cs:
```diff
@@ -31,6 +33,253 @@ public static IEndpointConventionBuilder MapRazorPages(this IEndpointRouteBuilde
+        /// result in an HTTP 404.
+        /// </para>
+        /// <para>
+        /// The order of the registered endpoint will be <c>int.MaxValue</c>.
```
Where does this get set? I didn't find any references to the order being set in the code.

rynowak replied:
`MapFallback` sets the order

pranavkm commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,126 @@
+                {
+                    // If there's no match this is a configuration error. We can't really check
+                    // during startup that the action you configured exists.
+                    throw new InvalidOperationException(
```
Use the resx luke

rynowak replied:
To what end? we have no plans to add localization. It's just homework.

pranavkm commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointMatcherPolicy.cs:
```diff
@@ -0,0 +1,126 @@
+    internal class DynamicPageEndpointMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
+    {
+        private readonly DynamicPageEndpointSelector _selector;
+        private readonly IPageLoader _loader;
```
This will complain once I get my changes in to make it Obsolete

rynowak replied:
Yes.

pranavkm commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointSelector.cs:
```diff
@@ -0,0 +1,56 @@
+
+        private ActionSelectionTable<RouteEndpoint> Table => _cache.EnsureInitialized();
+
+        // This is async because the page version will be async. We don't want to put ourselves in a
```
> This is async because the page version will be async.

This is the page version. Neither the controller selector nor the page do async work. Should they be synchronous?

rynowak replied:
I'll remove the comment - also this will be async once I use your new thing.

 I **insist** that these be asynchronous because I anticipate that they will be public. I do not want them to be made synchronous and then proliferate usage all over the world.

rynowak replied:
GAH. You are the correct. These are no longer async.

pranavkm commented on src/Mvc/Mvc.Core/src/Infrastructure/StringArrayComparer.cs:
```diff
@@ -0,0 +1,79 @@
+                return 0;
+            }
+
+            var hash = new HashCodeCombiner();
```
Can you use `System.HashCode` instead?

pranavkm commented on src/Http/Routing/src/DataSourceDependentCache.cs:
```diff
@@ -8,6 +8,8 @@
 
 namespace Microsoft.AspNetCore.Routing
 {
+    // FYI: This class is also linked into MVC. If you make changes to the API you will
```
You kinda don't need this comment per se. The build would fail if you #yolo change this.

rynowak replied:
meh.

Commit:
Make StaticFiles Noop for middleware

Teaches all of the static files middleware (incl default files,
directory browser) to noop when an endpoint is selected. This is
desirable so you can place them after routing if you want with no ill
effect.

rynowak commented:
Alright this is updated and I've addressed all of the feedback. @Tratcher - there is a new commit with your name on it.

Tratcher commented on src/Middleware/StaticFiles/src/LoggerExtensions.cs:
```diff
@@ -75,6 +76,10 @@ static LoggerExtensions()
+            _endpointMatched = LoggerMessage.Define(
+                logLevel: LogLevel.Debug,
+                eventId: new EventId(15, "EndpointMatched"),
+                formatString: "The request matched an endpoint.");
```
"Static files was skipped as the request already matched an endpoint."

Tratcher commented on src/Middleware/StaticFiles/src/StaticFileMiddleware.cs:
```diff
@@ -76,6 +76,10 @@ public async Task Invoke(HttpContext context)
             {
                 _logger.RequestMethodNotSupported(context.Request.Method);
             }
+            else if (!fileContext.ValidateNoEndpoint())
```
Move first.

rynowak commented:
@aspnet/brt - Come on Bort! It's time to leave. Sorry, are you talking to me? No, my son is also named Bort.

anurse commented:
Mop and Bucket man to the exit of the Nauseator!

This is https://github.com/aspnet/AspNetCore/issues/6597 which is **supposedly** to be fixed "soon" (hence I wasn't as aggressive about skipping). Seems @halter73 and @BrennanConroy are debating the right way to fix it though so I'll add a skippy-skip.

rynowak commented:
> which is supposedly to be fixed "soon" (hence I wasn't as aggressive about skipping)

The only good test is a skipped test!

rynowak commented:
@aspnet/brt @HaoK - can you take a look at this #helixfail
